packages/pubsub/src/connection-pool.js

```diff
@@ -0,0 +1,190 @@
+/*!
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module pubsub/connectionPool
+ */
+
+'use strict';
+
+var arrify = require('arrify');
+var common = require('@google-cloud/common');
+var each = require('async-each');
+var events = require('events');
+var extend = require('extend');
+var is = require('is');
+var util = require('util');
+var uuid = require('uuid');
+
+/**
+ * @type {module:pubsub/message}
+ * @private
+ */
+var Message = require('./message.js');
+
+/**
+ *
+ */
+function ConnectionPool(subscription, options) {
+  this.subscription = subscription;
+  this.connections = new Map();
+  this.isPaused = false;
+  this.isOpen = false;
+
+  this.settings = extend({
+    maxConnections: 5,
+    ackDeadline: 10000
+  }, options);
+
+  events.EventEmitter.call(this);
+
+  this.open();
+}
+
+util.inherits(ConnectionPool, events.EventEmitter);
+
+/**
+ *
+ */
+ConnectionPool.prototype.acquire = function(id, callback) {
```

Comment:
Sorry, I've been in the habit of writing documentation after the code gets approved, but I can start documenting as I'm coding from now on.

One of the requirements for using multiple connections is to make sure that any messages that are ack'd/nack'd are done so on the same connection they were received on. I was thinking it might be better to not store a pointer to the connection itself since it could potentially close and storing it would prevent it from being garbage collected. So I basically assigned each connection created an id that was stored on the message instead and this method attempted to retrieve said connection and if it failed to do so it would just grab the first available connection.